### PR TITLE
Add metrics backends

### PR DESCRIPTION
**Pull request by mmihic**
Graphite + StackDriver + web endpoint exposing metrics as JSON over HTTP
for manual expection and pull-based tooling such as Prometheus

### CODE REVIEW

File: misk/src/main/kotlin/misk/metrics/backends/MetricsBackendConfig.kt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package misk.metrics.backends
+
+import misk.config.Config
+import misk.metrics.backends.graphite.GraphiteBackendConfig
+import misk.metrics.backends.stackdriver.StackDriverBackendConfig
+
+data class MetricsBackendConfig(
+        val graphite: GraphiteBackendConfig?,
+        val stack_driver: StackDriverBackendConfig?) : Config
```
Comment (JGulbronson) [inline misk/src/main/kotlin/misk/metrics/backends/MetricsBackendConfig.kt]: nit: move `): Config` to a newline, to give each field its own line.

File: misk/src/main/kotlin/misk/metrics/backends/graphite/GraphiteReporterService.kt [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+import com.google.common.util.concurrent.AbstractIdleService
+import java.util.concurrent.TimeUnit
+import javax.inject.Inject
+
+internal class GraphiteReporterService @Inject internal constructor(
+        private val config: GraphiteBackendConfig,
+        private val reporter: GraphiteReporter
+) : AbstractIdleService() {
+    override fun startUp() = reporter.start(config.interval.toMillis(), TimeUnit.MILLISECONDS)
+    override fun shutDown() = reporter.stop()
```
Comment (JGulbronson) [inline misk/src/main/kotlin/misk/metrics/backends/graphite/GraphiteReporterService.kt]: This is awesome how short these are.

File: misk/src/main/kotlin/misk/metrics/backends/stackdriver/StackDriverBackendConfig.kt [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+import misk.config.Config
+import java.time.Duration
+
+data class StackDriverBackendConfig(
+        val project_id: String,
+        val service_path: String,
+        val root_url: String,
+        val interval: Duration,
+        val batch_size: Int = 200) : Config
```
Comment (JGulbronson) [inline misk/src/main/kotlin/misk/metrics/backends/stackdriver/StackDriverBackendConfig.kt]: nit: give each field its own line and move the `) : Config` bit to its own line.

File: misk/src/main/kotlin/misk/metrics/backends/stackdriver/StackDriverBackendModule.kt [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import com.google.inject.Singleton
+import misk.config.AppName
+import misk.inject.addMultibinderBinding
+import misk.inject.to
+
+class StackDriverBackendModule : AbstractModule() {
+    override fun configure() {
+        bind(StackDriverSender::class.java)
+                .to(StackDriverBatchedSender::class.java)
+                .`in`(Singleton::class.java)
```
Comment (JGulbronson) [inline misk/src/main/kotlin/misk/metrics/backends/stackdriver/StackDriverBackendModule.kt]: You can use `inSingleton()` which we define in `Guice.kt` to avoid having the ugly quotes.

File: misk/src/main/kotlin/misk/metrics/backends/stackdriver/StackDriverBatchedSender.kt [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    class ResponseCallback : JsonBatchCallback<Empty>() {
+        override fun onFailure(e: GoogleJsonError?, responseHeaders: HttpHeaders?) {
+            log.warn("failed to send timeseries to stack driver ${e?.toPrettyString()}")
+        }
+
+        override fun onSuccess(t: Empty?, responseHeaders: HttpHeaders?) {}
+    }
+
```
Comment (JGulbronson) [inline misk/src/main/kotlin/misk/metrics/backends/stackdriver/StackDriverBatchedSender.kt]: nit: extra line.

File: build.gradle [hunk 6]
```diff
@@ -24,6 +24,9 @@ ext {
           jacksonKotlin: 'com.fasterxml.jackson.module:jackson-module-kotlin:2.9.2',
           metricsParent: 'io.dropwizard.metrics:metrics-parent:3.1.0',
           metricsCore: 'io.dropwizard.metrics:metrics-core:3.1.0',
+          metricsGraphite: 'io.dropwizard.metrics:metrics-graphite:3.1.0',
+          metricsStackDriver: 'com.google.apis:google-api-services-monitoring:v3-rev425-1.23.0',
+          mockitoCore: "org.mockito:mockito-core:2.+",
```
Comment (JGulbronson) [inline build.gradle]: Is this actually being used? I don't see any `mockito` imports in the code.

File: misk/build.gradle [hunk 7]
```diff
@@ -34,9 +34,12 @@ dependencies {
   compile dep.jacksonKotlin
   compile dep.metricsCore
   compile dep.metricsParent
+  compile dep.metricsGraphite
+  compile dep.metricsStackDriver
 
   testCompile dep.junit
   testCompile dep.truth
   testCompile dep.loggingImpl
+  testCompile dep.mockitoCore
```
Comment (JGulbronson) [inline misk/build.gradle]: Doesn't seem to be used.

File: misk/src/test/kotlin/misk/metrics/backends/stackdriver/StackDriverReporterTest.kt [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+import java.math.BigDecimal
+import java.math.BigInteger
+import java.util.concurrent.TimeUnit
+import javax.inject.Inject
+import javax.inject.Singleton
+
+internal class StackDriverReporterTest {
+    companion object {
+        @JvmStatic
+        val APP_NAME = "my_app"
```
Comment (JGulbronson) [inline misk/src/test/kotlin/misk/metrics/backends/stackdriver/StackDriverReporterTest.kt]: If these are just being used to test the `StackDriverReporter` class, they shouldn't need `@JvmStatic`.

File: misk/src/main/kotlin/misk/metrics/backends/stackdriver/StackDriverReporter.kt [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+        sender.send(timeSeries)
+    }
+
+    enum class MetricKind {
+        GAUGE,
+        CUMULATIVE
+    }
+
+    companion object {
+        @JvmStatic
```
Comment (JGulbronson) [inline misk/src/main/kotlin/misk/metrics/backends/stackdriver/StackDriverReporter.kt]: Since this is an internal class, I don't think you need `@JvmStatic`.

File: misk/src/main/kotlin/misk/metrics/backends/stackdriver/StackDriverBatchedSender.kt [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+            val request = CreateTimeSeriesRequest().setTimeSeries(timeSeries)
+            val create = timeSeriesApi.create(config.project_id, request).execute()
+            if (!create.isEmpty()) {
+                log.warn(create.toPrettyString())
+            }
+        }
+    }
+
+    class ResponseCallback : JsonBatchCallback<Empty>() {
+        override fun onFailure(e: GoogleJsonError?, responseHeaders: HttpHeaders?) {
```
Comment (JGulbronson) [inline misk/src/main/kotlin/misk/metrics/backends/stackdriver/StackDriverBatchedSender.kt]: Will `e` ever be nullable? I think sometimes Intellij marks things as nullable to be extra safe, but I don't think you'll get a failure with a null error here.

File: misk/src/test/kotlin/misk/metrics/backends/stackdriver/StackDriverReporterTest.kt [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+                metrics.gauges,
+                metrics.counters,
+                metrics.histograms,
+                null,
+                metrics.timers,
+                START_TIME, NOW)
+
+        val names = timeSeries.map { it.metric.type to it }.toMap()
+        assertThat(names.keys).containsExactly(
+                "custom.googleapis.com/dw/my_app/us-west2/165.33.45.25/im-basic/value",
```
Comment (JGulbronson) [inline misk/src/test/kotlin/misk/metrics/backends/stackdriver/StackDriverReporterTest.kt]: Are these coming from an actual HTTP request?

File: exemplar/src/main/kotlin/com/squareup/exemplar/ExemplarModule.kt [hunk 12]
```diff
@@ -15,5 +16,6 @@ class ExemplarModule : AbstractModule() {
         install(WebActionModule.create<ReadinessCheckAction>())
         install(WebActionModule.create<LivenessCheckAction>())
         install(WebActionModule.create<NotFoundAction>())
+        install(WebActionModule.create<MetricsAction>())
```
Comment (JGulbronson) [inline exemplar/src/main/kotlin/com/squareup/exemplar/ExemplarModule.kt]: This should be before `NotFoundAction`. Maybe we should warn if `NotFoundAction` isn't last?

File: exemplar/src/main/java/com/squareup/exemplar/ExemplarJavaModule.java [hunk 13]
```diff
@@ -10,5 +11,6 @@
     install(WebActionModule.create(HelloJavaAction.class));
     install(WebActionModule.create(InternalErrorAction.class));
     install(WebActionModule.create(NotFoundAction.class));
+    install(WebActionModule.create(MetricsAction.class));
```
Comment (JGulbronson) [inline exemplar/src/main/java/com/squareup/exemplar/ExemplarJavaModule.java]: Same here, should be one line up.

**Comment by JGulbronson** [comment]
Outside the scope of this PR, but we might want to look at allowing endpoints to be exposed using one of the Prometheus exposition formats (probably the text format to start), since I don't think it will understand the JSON.

File: misk/src/main/kotlin/misk/metrics/MetricsConfig.kt [hunk 14]
```diff
@@ -0,0 +1,7 @@
+package misk.metrics
+
+import misk.config.Config
+import misk.metrics.backends.MetricsBackendConfig
+
+data class MetricsConfig(val backends: MetricsBackendConfig) : Config
```
Comment (JGulbronson) [inline misk/src/main/kotlin/misk/metrics/MetricsConfig.kt]: This isn't used right now, should config be set up in a different PR?

File: misk/src/main/kotlin/misk/metrics/backends/MetricsBackendConfig.kt [hunk 15]
```diff
@@ -0,0 +1,9 @@
+package misk.metrics.backends
+
+import misk.config.Config
+import misk.metrics.backends.graphite.GraphiteBackendConfig
+import misk.metrics.backends.stackdriver.StackDriverBackendConfig
+
+data class MetricsBackendConfig(
```
Comment (JGulbronson) [inline misk/src/main/kotlin/misk/metrics/backends/MetricsBackendConfig.kt]: This only seems to be used in `MetricsConfig`, which isn't used. Should this be moved to a separate PR?

File: misk/src/test/kotlin/misk/metrics/backends/stackdriver/StackDriverReporterTest.kt (hunk 11 above)
Comment (mmihic) [reply misk/src/test/kotlin/misk/metrics/backends/stackdriver/StackDriverReporterTest.kt]: Stackdriver uses path format to identify metrics. The custom.googleapis.com/dw prefix is required to identify these as custom metrics.

File: exemplar/src/main/kotlin/com/squareup/exemplar/ExemplarModule.kt (hunk 12 above)
Comment (mmihic) [reply exemplar/src/main/kotlin/com/squareup/exemplar/ExemplarModule.kt]: Yes, seem fragile for binding statements to have an implicit required ordering. I'd hard error if the required ordering is not correct. Same for interceptors.

File: misk/src/main/kotlin/misk/metrics/web/MetricsAction.kt [hunk 16]
```diff
@@ -0,0 +1,14 @@
+package misk.metrics.web
+
+import misk.metrics.Metrics
+import misk.web.Get
+import misk.web.JsonResponseBody
+import misk.web.actions.WebAction
+import javax.inject.Inject
+
+/** Exposes metrics as JSON over HTTP */
+class MetricsAction @Inject internal constructor(val metrics: Metrics) : WebAction {
```
[the remark this answers is not in the document]
Comment (mmihic) [reply misk/src/main/kotlin/misk/metrics/web/MetricsAction.kt]: Sure, although doesn't really matter.

File: misk/src/main/kotlin/misk/metrics/MetricsConfig.kt (hunk 14 above)
Comment (mmihic) [reply misk/src/main/kotlin/misk/metrics/MetricsConfig.kt]: The top-level config is the top-level container for the back-end specific configs introduced in this PR.  IMO it makes the most sense to add all of that configuration in this PR.  We don't use the metrics config in exemplar because I don't think we want to have an exemplar configuration specifying a specific metrics backend, with associated port, project, etc.

File: misk/src/main/kotlin/misk/metrics/backends/MetricsBackendConfig.kt (hunk 15 above)
Comment (mmihic) [reply misk/src/main/kotlin/misk/metrics/backends/MetricsBackendConfig.kt]: Same comment as above

**Comment by mmihic** [comment]
I'd like to have content-type dispatching to add a plain-text endpoint for metrics.

File: exemplar/src/main/java/com/squareup/exemplar/ExemplarJavaModule.java (hunk 13 above)
Comment (mmihic) [reply exemplar/src/main/java/com/squareup/exemplar/ExemplarJavaModule.java]: Done. BTW the java module is missing the liveness and readiness actions.

File: build.gradle (hunk 6 above)
Comment (mmihic) [reply build.gradle]: Removed. I was originally trying to mock stackdriver to test the actual batch sending code, but that turns out to be impossible to do a meaningful way given the structure of the client.
